src/vse_sync_pp/analyzers/pmc.py

```diff
@@ -0,0 +1,147 @@
+    """Transition from the holdover out of spec state"""
+    try:
+        clock_state_analyzer.state = HOLDOVER_OUT_OF_SPEC_TRANSITION[clock_class]
+    except KeyError:
+        return False
+
+STATE_TRANSITION = {
+    STATE_FREERUN: handle_freerun_transition,
+    STATE_LOCKED: handle_locked_transition,
+    STATE_HOLDOVER_IN_SPEC: handle_holdover_in_spec_transition,
+    STATE_HOLDOVER_OUT_OF_SPEC1: handle_holdover_out_of_spec_transition,
+    STATE_HOLDOVER_OUT_OF_SPEC2: handle_holdover_out_of_spec_transition,
+    STATE_HOLDOVER_OUT_OF_SPEC3: handle_holdover_out_of_spec_transition,
+}
+
+FREERUN_TRANSITION = {
+    STATE_FREERUN: STATE_FREERUN,
+    STATE_LOCKED: STATE_LOCKED,
+}
+
+LOCKED_TRANSITION = {
+    STATE_LOCKED: STATE_LOCKED,
+    STATE_HOLDOVER_IN_SPEC: STATE_HOLDOVER_IN_SPEC,
+}
+
+HOLDOVER_IN_SPEC_TRANSITION = {
+    STATE_LOCKED: STATE_LOCKED,
+    STATE_HOLDOVER_IN_SPEC: STATE_HOLDOVER_IN_SPEC,
+    STATE_HOLDOVER_OUT_OF_SPEC1: STATE_HOLDOVER_OUT_OF_SPEC1,
+    STATE_HOLDOVER_OUT_OF_SPEC2: STATE_HOLDOVER_OUT_OF_SPEC2,
+    STATE_HOLDOVER_OUT_OF_SPEC3: STATE_HOLDOVER_OUT_OF_SPEC3,
+}
+
+HOLDOVER_OUT_OF_SPEC_TRANSITION = {
+    STATE_LOCKED: STATE_LOCKED,
+    STATE_HOLDOVER_OUT_OF_SPEC1: STATE_HOLDOVER_OUT_OF_SPEC1,
+    STATE_HOLDOVER_OUT_OF_SPEC2: STATE_HOLDOVER_OUT_OF_SPEC2,
+    STATE_HOLDOVER_OUT_OF_SPEC3: STATE_HOLDOVER_OUT_OF_SPEC3,
+}
+
+class ClockStateAnalyzer(Analyzer):
+    """Analyze clock state
+    """
+    id_ = 'phc/gm-settings'
+    parser = id_
+    def __init__(self, config):
+        super().__init__(config)
+        # minimum test duration for a valid test
+        self._duration = config.parameter('min-test-duration/s')
+        self.state = None
+        self.transition_count = 0
+        self.clock_class_count = {
+            STATE_FREERUN: 0,
+            STATE_LOCKED: 0,
+            STATE_HOLDOVER_IN_SPEC: 0,
+            STATE_HOLDOVER_OUT_OF_SPEC1: 0,
+            STATE_HOLDOVER_OUT_OF_SPEC2: 0,
+            STATE_HOLDOVER_OUT_OF_SPEC3: 0,
+        }
+    def prepare(self, rows):
+        idx = 0
+        try:
+            tstart = rows[0].timestamp
+        except IndexError:
+            pass
+        else:
+            while idx < len(rows):
+                if tstart <= rows[idx].timestamp:
+                    break
+                idx += 1
+        return super().prepare(rows[idx:])
+    def test(self, data):
+        if len(data) == 0:
+            return (False, "no data")
+        
+        for index, row in data.iterrows():
+            clock_class = row['clock_class']
+ 
+            if (self.state is None) and (clock_class in STATE_TRANSITION):
+                self.state = clock_class
+            else:
+                if clock_class != self.state:
+                    self.transition_count += 1
+                
+                try:
+                    transitioned = STATE_TRANSITION[self.state](self, clock_class)
```

Comment:
this obscures the fact your changing the state in self. Just check if its a valid state and then update it.

Reply:
Good point. Done